Google.Cloud.Spanner.NHibernate.Tests/NHibernateMockServerTests.cs

```diff
@@ -0,0 +1,139 @@
+                    Assert.Equal(Value.ForString("Pete"), request.Params.Fields["p1"]);
+                    Assert.Equal(Value.ForString("Allison"), request.Params.Fields["p2"]);
+                    Assert.Equal(Value.ForString("1"), request.Params.Fields["p3"]);
+                }
+            );
+
+            _fixture.SpannerMock.Reset();
+            // Update the singer and verify that the version number is included in the WHERE clause and is updated.
+            var updateSql = "UPDATE SingerWithVersion SET Version = @p0, FirstName = @p1, LastName = @p2 WHERE SingerId = @p3 AND Version = @p4";
+            var updateAlbumSql = "UPDATE AlbumWithVersion SET SingerWithVersion = null WHERE SingerWithVersion = @p0";
+            _fixture.SpannerMock.AddOrUpdateStatementResult(updateSql, StatementResult.CreateUpdateCount(1L));
+            _fixture.SpannerMock.AddOrUpdateStatementResult(updateAlbumSql, StatementResult.CreateUpdateCount(1L));
+
+            singer.LastName = "Peterson - Allison";
+
+            using (var session = _fixture.SessionFactory.OpenSession())
+            {
+                using (var transaction = session.BeginTransaction())
+                {
+                    await session.SaveOrUpdateAsync(singer);
+                    await session.FlushAsync();
+                    await transaction.CommitAsync();
+                }
+            }
+
+            var updateQueryRequest = _fixture.SpannerMock.Requests.OfType<ExecuteSqlRequest>().FirstOrDefault();
+            Assert.NotNull(updateQueryRequest);
+
+            Assert.Equal(updateSql, updateQueryRequest.Sql);
+            // Verify that the version that is set is 2.
+            Assert.Equal(Value.ForString("2"), updateQueryRequest.Params.Fields["p0"]);
+
+            Assert.Equal(Value.ForString("Pete"), updateQueryRequest.Params.Fields["p1"]);
+            Assert.Equal(Value.ForString("Peterson - Allison"), updateQueryRequest.Params.Fields["p2"]);
+            Assert.Equal(Value.ForString("1"), updateQueryRequest.Params.Fields["p3"]);
+            // Verify that the version that is checked is 1.
+            Assert.Equal(Value.ForString("1"), updateQueryRequest.Params.Fields["p4"]);
+        }
+
+        [Fact]
+        public async Task UpdateFailsIfVersionNumberChanged()
+        {
+            var updateSql = "UPDATE SingerWithVersion SET Version = @p0, FirstName = @p1, LastName = @p2 WHERE SingerId = @p3 AND Version = @p4";
+            var selectSql = "SELECT singerwith0_.SingerId as singerid1_2_0_, singerwith0_.Version as version2_2_0_, singerwith0_.FirstName as firstname3_2_0_, singerwith0_.LastName as lastname4_2_0_ FROM SingerWithVersion singerwith0_ WHERE singerwith0_.SingerId=@p0";
+            // Set the update count to 0 to indicate that the row was not found.
+            var insertSql = "INSERT INTO SingerWithVersion (Version, FirstName, LastName, SingerId) VALUES (@p0, @p1, @p2, @p3)";
```

Comment:
This test should not generate an `INSERT INTO ...` statement, and also an insert statement should never return an update count of zero (unless it uses a `SELECT` statement that returns zero rows). The problem is caused by the fact that `session.Merge(..)` is not the same as `DatabaseContext.Attach` in Entity Framework. `Merge` will check whether the row actually exists in the database, and if not, insert it into the database and then attach it to the session. `Attach` will only attach it to the context and assume that it exists in the database. There is (as far as I know) no `Attach` or equivalent method in NHibernate, so the test needs to be set up in a slightly different way:

1. Instead of attaching the entity to the session manually, you need to load it. For that you need to add a result to the mock server for the load operation:

```csharp
            _fixture.SpannerMock.AddOrUpdateStatementResult(selectSql, StatementResult.CreateResultSet(new List<Tuple<V1.TypeCode, string>>
            {
                Tuple.Create(V1.TypeCode.Int64, "singerid1_2_0_"),
                Tuple.Create(V1.TypeCode.Int64, "version2_2_0_"),
                Tuple.Create(V1.TypeCode.String, "firstname3_2_0_"),
                Tuple.Create(V1.TypeCode.String, "lastname4_2_0_")
            }, new List<object[]>
            {
                new object[]{"1", "1", "Pete", "Allison"}
            }));
```

And then remove the insert statement, the `new SingerWithVersion { ... }` and the `session.Merge(..)` calls and replace it with a simple load call:

```csharp
                using (var transaction = session.BeginTransaction())
                {
                    var singer = await session.LoadAsync<SingerWithVersion>(1L);
                    singer.LastName = "Allison - Peterson";
                    await Assert.ThrowsAsync<StaleStateException>(() => session.FlushAsync());
                    // Update the update count to 1 to simulate a resolved version conflict.
                    _fixture.SpannerMock.AddOrUpdateStatementResult(updateSql, StatementResult.CreateUpdateCount(1L));
                    await session.FlushAsync();
                    await transaction.CommitAsync();
                }
```